### DISCUSSION TITLE
One to one mapping

### DISCUSSION THREAD

**Discussion by pintariching** [diesel-rs/diesel#3152] [Q&A] [answered]
Let's say I have two structs like so:
```
pub struct Machine {
  pub id: i64,
  pub name: String,
  pub manufacturer_id: i64
  pub manufacturer: Manufacturer
}

pub struct Manufacturer {
  pub id: i64,
  pub name: String,
}
```

Is it possible to write a simple single query to map these structs so I can use `serde` return a single JSON object like so:
```
{
  "machine": {
    "id": 1,
    "name": "Station 1",
    "manufacturer_id": 4,
    "manufacturer": {
      "id": 4,
      "name": "Company ltd."  
    }
  }
}
```

**Comment by weiznich** [comment, marked as answer]
Yes that's possible using a join + a custom select clause. Assuming both structs implement `Queryable` the following code should work:

```rust
// For reference the corresponding table definitions

table! {
    machines (id) {
        id -> Integer,
        name -> Text,
        manufacturer_id -> Integer,
    }
}

table! {
    manufacturers (id) {
        id -> Integer,
         name -> Text,
    }
}

joinable!(machines -> manufacturers (manufacturer_id));

// use the following query to receive the data in the correct shape:

machines::table.inner_join(manufacturers::table).select((machines::id, machines::name, machines::manufacturer_id, manufacturers::all_columns).load::<Machine>(&conn)?;

```

**Reply by pintariching** [reply]
Okay this worked perfectly! I had to watch out for the column order to match exactly as it was defined in the struct and the schema definition. One thing I don't really like is that in the `.select()` you have to verbosely list all of the columns of the `machine` struct and just pass in the `manufacturer::all_columns` for the, I guess, "child" struct? If the "parent" struct has a lot of columns, is there no way of using a statement like `.select((machine::all_columns, manufacturer::all_columns))`?

**Reply by weiznich** [reply]
The upcoming 2.0 release introduces a [`Selectable`](https://docs.diesel.rs/master/diesel/expression/trait.Selectable.html) trait that simplifies the construction of such select clauses.

**Reply by pintariching** [reply]
Okay that's very cool! Thank you very much for the information!
